pkg/controller/security/security_manager.go

```diff
@@ -0,0 +1,256 @@
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package security
+
+import (
+	"sync"
+	"time"
+
+	"container/heap"
+
+	istiosecurity "istio.io/istio/pkg/security"
+
+	"kmesh.net/kmesh/pkg/bpf"
+	"kmesh.net/kmesh/pkg/constants"
+	"kmesh.net/kmesh/pkg/controller/workload"
+	"kmesh.net/kmesh/pkg/controller/workload/cache"
+	"kmesh.net/kmesh/pkg/logger"
+	"kmesh.net/kmesh/pkg/nets"
+)
+	
+type certExp struct {
+	hostName string
+	exp time.Time
+	removed bool
+}
+
+type certState struct {
+	count uint32
+	certExp *certExp
+}
+
+type PriorityQueue []*certExp
+
+var (
+    certsMaps sync.Map
+	certsStates sync.Map
+	ScClient SecretManager
+    log = logger.NewLoggerField("security")	
+)
+
+type SecretManager struct {
+	caClient *KmeshCaClient
+
+	// configOptions includes all configurable params for the cache.
+	configOptions *istiosecurity.Options
+
+	// storing certificates
+	certsCache *sync.Map
+
+	// certificates states
+	certsStates *sync.Map
+
+	//caRootPath string
+	pending	PriorityQueue
+}
+   
+func (pq PriorityQueue) Len() int { return len(pq) }
+
+func (pq PriorityQueue) Less(i, j int) bool {
+	return pq[i].exp.Before(pq[j].exp)
+}
+
+func (pq PriorityQueue) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+}
+
+func (pq *PriorityQueue) Push(x interface{}) {
+	item := x.(*certExp)
+	*pq = append(*pq, item)
+}
+
+func (pq *PriorityQueue) Pop() interface{} {
+	old := *pq
+	n := len(old)
+	item := old[n-1]
+	*pq = old[0 : n-1]
+	return item
+}
+
+// NewSecretManager creates a new SecretManager.
+func NewSecretManager(bpfWorkloadObj *bpf.BpfKmeshWorkload) error {
+
+	tlsOpts = &TLSOptions{
+		RootCert:      constants.RootCertPath,
+	}
+
+	options:= NewSecurityOptions()
+	caClient, err := NewKmeshCaClient(options, tlsOpts)
+	if err != nil {
+		return err
+	}
+	
+	pq := make(PriorityQueue, 0)
+	heap.Init(&pq)
+
+	ScClient= SecretManager{
+		caClient:      caClient,
+		configOptions: options,
+		certsCache: &certsMaps,
+		certsStates: &certsStates,
+		pending:	pq,
+	}
+	go ScClient.refreshExpireCerts()
+	go ScClient.updateCerts(bpfWorkloadObj)
+	return nil
+}
+ 
+func GetSecretManager() *SecretManager {
+	return &ScClient
+}
+	
+// Automatically check and refresh when the validity period expires
+func (sc *SecretManager) refreshExpireCerts() {
+	for {
+		expireTime := time.Now().Add(15 * time.Minute)
+		if (len(sc.pending) > 0) {
+			expireTime = sc.pending[0].exp
+		}
+		select{
+		case <-time.After(time.Until(expireTime.Add(-5 * time.Minute))):
+			next := heap.Pop(&sc.pending).(*certExp)
+			if next.removed == true {
```

Comment:
how do you change the element that has been pushed into the heap without lock,

Maybe need to rethink about how to know whether a sa related workload has been deleted

Reply:
yes，The removed field is now removed and the pending priority queue with lock is redesigned.